nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-shell-authorizer/src/main/java/org/apache/nifi/authorization/ShellUserGroupProvider.java

```diff
@@ -0,0 +1,367 @@
+                if (record.length > 1) {
+                    String name = record[0],
+                        id = record[1];
+                    if (name != null && id != null && !name.equals("") && !id.equals("")) {
+                        User user = new User.Builder().identity(name).identifier(id).build();
+                        byId.put(id, user);
+                        byName.put(name, user);
+                        logger.debug("refreshed user: " + user);
+                    } else {
+                        logger.warn("null or empty user name: " + name + " or id: " + id);
+                    }
+                }
+            });
+
+        synchronized (usersById) {
+            usersById.clear();
+            usersById.putAll(byId);
+        }
+
+        synchronized (usersByName) {
+            usersByName.clear();
+            usersByName.putAll(byName);
+            logger.debug("refreshUsers users now size: " + usersByName.size());
+        }
+    }
+
+    private void refreshGroups() {
+        Map<String, Group> groups = new HashMap<>();
+        List<String> lines;
+
+        try {
+            lines = ShellRunner.runShell(selectedShellCommands.getGroupsList());
+        } catch (final IOException ioexc) {
+            logger.error("refreshGroups list groups shell exception: " + ioexc);
+            return;
+        }
+
+        lines.forEach(line -> {
+                String[] record = line.split(":");
+                if (record.length > 1) {
+                    Set<String> users = new HashSet<>();
+                    String name = record[0], id = record[1];
+
+                    try {
+                        List<String> userLines = ShellRunner.runShell(String.format(selectedShellCommands.getGroupMembers(), name));
+                        if (userLines.size() > 0) {
+                            users.addAll(Arrays.asList(userLines.get(0).split(",")));
+                        }
+                    } catch (final IOException ioexc) {
+                        logger.error("refreshGroups list membership shell exception: " + ioexc);
+                    }
+
+                    if (name != null && id != null && !name.equals("") && !id.equals("")) {
+                        Group group = new Group.Builder().name(name).identifier(id).addUsers(users).build();
+                        groups.put(id, group);
+                        logger.debug("refreshed group: " + group);
+                    } else {
+                        logger.warn("null or empty group name: " + name + " or id: " + id);
+                    }
+                }
+            });
+
+        List<String> sublines;
+        try {
+            sublines = ShellRunner.runShell(selectedShellCommands.getUsersList());
+        } catch (final IOException ioexc) {
+            logger.error("refreshGroups list groups shell exception: " + ioexc);
+            return;
+        }
+
+        sublines.forEach(line -> {
```

Comment:
I think this block should be refactored to a component method which encapsulates the behavior of iterating over each _user_ and returning a collection of `Group` objects, each fully populated with their respective _user_ `String`s.

Reply:
Agreed.  Change inbound.